python/taichi/lang/_wrap_inspect.py

```diff
@@ -0,0 +1,143 @@
+import inspect
+import os
+import tempfile
+
+import sourceinspect
+from taichi._logging import warn
+
+_builtin_getfile = inspect.getfile
+_builtin_findsource = inspect.findsource
+
+
+def _find_source_with_custom_getfile_func(func, obj):
+    inspect.getfile = func  # replace with our custom func
+    source = inspect.findsource(obj)
+    inspect.getfile = _builtin_getfile  # restore
+    return source
+
+
+def _blender_get_text_name(filename: str):
+    if filename.startswith(os.path.sep) and filename.count(os.path.sep) == 1:
+        return filename[1:]  # "/Text.001" --> "Text.001"
+
+    index = filename.rfind('.blend' + os.path.sep)
+    if index != -1:
+        return filename[index + 7:]  # "hello.blend/test.py" --> "test.py"
+
+    return None
+
+
+def _blender_findsource(obj):
+    try:
+        import bpy  # pylint: disable=import-outside-toplevel
+    except ImportError:
+        raise IOError('Not in Blender environment!')
+
+    filename = _builtin_getfile(obj)
+    text_name = _blender_get_text_name(filename)
+    if text_name is None:
+        raise IOError(
+            'Object `{obj.__name__}` is not defined in a .blend file!')
+
+    lines = bpy.data.texts[text_name].as_string()
+    # Now we have found the filename and code lines.
+    # We first check if they are already cached, to avoid file io in each query.
+    try:
+        filename = _blender_findsource._saved_inspect_cache[lines]  # pylint: disable=no-member
+    except KeyError:
+        fd, filename = tempfile.mkstemp(prefix='_Blender_',
```

Comment:
who's responsible for removing these temp files?

Reply:
Good catch... Updated to use `atexit.register` to remove the file.